Sources/DeepLearning/Operators/Basic.swift

```diff
@@ -0,0 +1,710 @@
+    }
+
+    /// Concatenates tensors along the specified axis.
+    /// - Precondition: The tensors must have the same dimensions, except for the
+    ///   specified axis.
+    /// - Precondition: The axis must be in the range `-rank..<rank`.
+    @inlinable
+    @differentiable(vjp: _vjpConcatenated where Scalar: TensorFlowFloatingPoint)
+    func concatenated(with other: Tensor, alongAxis axis: Int = 0) -> Tensor {
+        return Tensor(concatenating: [self, other], alongAxis: axis)
+    }
+
+    /// Concatenation operator.
+    /// - Note: `++` is a custom operator that does not exist in Swift, but does
+    ///   in Haskell/Scala. Its addition is not an insignificant language change
+    ///   and may be controversial. The existence/naming of `++` will be discussed
+    ///   during a later API design phase.
+    @inlinable
+    @differentiable(where Scalar: TensorFlowFloatingPoint)
+    static func ++ (lhs: Tensor, rhs: Tensor) -> Tensor {
+        return lhs.concatenated(with: rhs)
+    }
+}
+
+internal extension Tensor where Scalar: TensorFlowFloatingPoint {
+    @inlinable
+    func _vjpTransposed(
+        withPermutations permutations: Tensor<Int32>
+    ) -> (Tensor, (Tensor) -> Tensor) {
+        let value = transposed(withPermutations: permutations)
+        return (value, { $0.transposed(withPermutations: permutations) })
+    }
+
+    @inlinable
+    func _vjpTransposed(withPermutations permutations: [Int]) -> (Tensor, (Tensor) -> Tensor) {
+        let value = transposed(withPermutations: permutations)
+        return (value, { $0.transposed(withPermutations: permutations) })
+    }
+
+    @inlinable
+    func _vjpTransposed(withPermutations permutations: Int...) -> (Tensor, (Tensor) -> Tensor) {
+        let value = transposed(withPermutations: permutations)
+        return (value, { $0.transposed(withPermutations: permutations) })
+    }
+
+    @inlinable
+    func _vjpTransposed() -> (Tensor, (Tensor) -> Tensor) {
+        return (transposed(), { $0.transposed() })
+    }
+
+    @inlinable
+    func _vjpConcatenated(
+        with other: Tensor,
+        alongAxis axis: Int
+    ) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
+        let idx = axis < 0 ? axis + rank: axis
+        let splits = Tensor<Int32>([shapeTensor[idx], other.shapeTensor[idx]])
+        return (concatenated(with: other, alongAxis: axis), { result in
+            let gradients = result.split(sizes: splits, alongAxis: axis)
+            return (gradients[0], gradients[1])
+        })
+    }
+}
+
+//===------------------------------------------------------------------------------------------===//
+// Broadcasting
+//===------------------------------------------------------------------------------------------===//
+
+// TODO: What about precedence? Why is this operator used for broadcasting?
+infix operator .=
+
+public extension Tensor {
+    @inlinable
+    func broadcast(toShape shape: Tensor<Int32>) -> Tensor {
```

Comment:
BTW, here's a related TODO: [TF-430](https://bugs.swift.org/browse/TF-430). Need not be done in this patch though!

Reply:
Could you apply this to `unbroadcast(to:)` and `unbroadcast(toShape:)` as well?